keps/sig-cli/5091-kubectl-checkpoint/README.md

```diff
@@ -0,0 +1,1205 @@
+user stories would benefit from additional integration into Kubernetes,
+but most user stories can already be used today. Implementation of this
+KEP would make the user experience more pleasant. Especially the container
+migration user story has many possibilities for optimization. CRIU, which is on
+the lowest level of the checkpoint/restore stack, offers the possibility to
+decrease container downtime during migration by techniques well known in virtual
+machine migration like pre-copy or post-copy migration.
+
+#### Forensic Container Checkpointing
+
+To analyze unusual activities in a container, the container should
+be checkpointed without stopping the container or without the container
+knowing it was checkpointed. Using checkpointing it is possible to take
+a copy of a running container for forensic analysis. The container will
+continue to run without knowing a copy was created. This copy can then
+be restored in another (sandboxed) environment in the context of another
+container engine for detailed analysis of a possible attack.
+
+#### Fast Container Startup - Warm Start
+
+In addition to forensic analysis of a container checkpointing can be used to
+offer a way to quickly start containers. This is especially useful for
+containers that need a long time start. Either the software in the container
+needs a long time to initialize by loading many libraries or the container
+requires time to read data from a storage device. Using checkpointing it is
+possible to wait once until the container finished the initialization and save
+the initialized state to a checkpoint archive. Based on this checkpoint archive
+one or multiple copies of the container can be created without the need to wait
+for the initialization to finish. The startup time is reduced to the time
+necessary to read back all memory pages to their previous location.
+
+This feature is already used in production to decrease startup time of
+containers.
+
+Another similar use case for quicker starting containers has been reported in
+combination with persistent memory systems. The combination of checkpointed
+containers and persistent memory systems can reduce startup time after a reboot
+tremendously.
+
+#### Optimize Resource Utilization
+
+This use case is motivated by interactive long running containers. One very
+common problem with things like Jupyter notebooks or remote development
+environments is that users want to have them running for a very long time to be
+able to come back to them whenever needed. To not block unused resources while
+these containers are not used checkpointing enables the possibility to make a
+copy of these stateful workloads which can be restored when the user wants to
+continue this interactive workload.
+
+This use case is especially interesting for resources that are limited like
+GPUs.
+
+
+#### Container Migration
+
+One of the main use cases for checkpointing and restoring containers is
```

Comment:
migration between nodes? the IPs are most likely to be lost so the application has to be agnostic of the IP per example

Reply:
Added a paragraph concerning migration of TCP connections.